src/index.tsx

```diff
@@ -2,57 +2,168 @@ import React from 'react';
+    //   }
+    //   rowArray.push(<div className="board-row">{squareArray}</div>)
+    // }
+
+    return rowArray
+   
+  }
+  
 
+  render() {
     return (
       <div>
-        <div className="status">{status}</div>
-        <div className="board-row">
-          {this.renderSquare(0)}
-          {this.renderSquare(1)}
-          {this.renderSquare(2)}
-        </div>
-        <div className="board-row">
-          {this.renderSquare(3)}
-          {this.renderSquare(4)}
-          {this.renderSquare(5)}
-        </div>
-        <div className="board-row">
-          {this.renderSquare(6)}
-          {this.renderSquare(7)}
-          {this.renderSquare(8)}
-        </div>
+          {this.loopSquare()}
       </div>
     );
   }
 }
 
-class Game extends React.Component {
+function calculateWinner(squares: Array<any>){
+  const lines = [
+    [0, 1, 2],
+    [3, 4, 5],
+    [6, 7, 8],
+    [0, 3, 6],
+    [1, 4, 7],
+    [2, 5, 8],
+    [0, 4, 8],
+    [2, 4, 6],
+  ];
+  for (let i = 0; i < lines.length; i++) {
+    const [a, b, c] = lines[i];
+    if (squares[a] && squares[a] === squares[b] && squares[a] === squares[c]) {
+      return {letter: squares[a], lines: lines[i]} ;
+    }
+  }
+  return null;
+}
+
+interface GameState{xIsNext: Boolean, history: Array<any>, stepNumber: number, coordinates: Array<any>}
+class Game extends React.Component<{}, GameState> {
+  constructor(props: {}){
+    super(props);
+    this.state = {
+      history: [{
+        squares: Array(9).fill(null),
+      }],
+      stepNumber: 0,
+      xIsNext: true,
+      coordinates: [{
+        col: 0,
+        row: 0,
+      }]
+    }
+  }
+
+  handleClick(i: number) {
+    const history = this.state.history;
+    const current = history[this.state.stepNumber];
+    const squares = current.squares.slice();
+    const coordinates = this.state.coordinates
+    let y: number, x: number
+    if (calculateWinner(squares) || squares[i]) {
+      return;
+    }
+    squares[i] = this.state.xIsNext ? 'X' : 'O';
+
+    y = (i%3)+1
+    x = (Math.floor(i/3)+1)
+    this.setState({
+      history: history.concat([{
+        squares: squares,
+      }]),
+      stepNumber: history.length,
+      xIsNext: !this.state.xIsNext,
+      coordinates : ([{
+        col: y,
+        row: x,
+      }])
+    });
+  }
+
+  jumpTo(step: number) {
+    this.setState({
+      stepNumber: step,
+      xIsNext: (step % 2) === 0,
+    });
+  }
+
+
   render() {
+    const history = this.state.history.slice(0, this.state.stepNumber + 1);
+    const current = history[history.length - 1];
+    const winner = calculateWinner(current.squares);
+
+    const moves = history.map((step, move) => {
+      const desc = move ?
+        'Go to move #' + move :
+        'Go to game start';
+      return (
+        <li key={move}>
+          <button  style={ move==this.state.stepNumber ? { fontWeight: 'bold' } : { fontWeight: 'normal' }}  onClick={() => this.jumpTo(move)}>{desc}</button>
```

Comment:
@meep130 Unless you're completely sure, you shouldn't use `==` a much as possible. In JS, please use `===`.

It's because `==` does **implicit type casting**

```js
0 == '0' // true because == makes '0' into a Number
0 === '0' //false because === keeps '0' as a String
```

Reply:
@yapTim done, please check again